### PR TITLE
cmake: Make install prefix of sdl2_config more compatible with SDL2

### PR DESCRIPTION
**Pull request by madebr**
- Prepend all cmake options with `SDL2COMPAT_`
- No need for `FindSDL3.cmake`, all SDL3 installations ship `SDL3Config.cmake` or `sdl3-config.cmake`.
- Install `SDL2Config.cmake` config files, who create the same targets as SDL2
- Create `SDL2COMPAT_VENDOR_INFO` cmake option, to store a vendor string.
- Set indentation of CMake to 2 (same as SDL)
- Add `.editorconfig`, based on the one from the main SDL repo

```
prefix
├── bin
│   └── sdl2-config
├── include
│   └── SDL2
│       ├── begin_code.h
│       ├── close_code.h
│       ├── SDL_assert.h
│       ├── SDL_atomic.h
│       ├── SDL_audio.h
│       ├── SDL_bits.h
│       ├── SDL_blendmode.h
│       ├── SDL_clipboard.h
│       ├── SDL_config_android.h
│       ├── SDL_config_emscripten.h
│       ├── SDL_config.h
│       ├── SDL_config_iphoneos.h
│       ├── SDL_config_macosx.h
│       ├── SDL_config_minimal.h
│       ├── SDL_config_ngage.h
│       ├── SDL_config_os2.h
│       ├── SDL_config_pandora.h
│       ├── SDL_config_windows.h
│       ├── SDL_config_wingdk.h
│       ├── SDL_config_winrt.h
│       ├── SDL_config_xbox.h
│       ├── SDL_copying.h
│       ├── SDL_cpuinfo.h
│       ├── SDL_egl.h
│       ├── SDL_endian.h
│       ├── SDL_error.h
│       ├── SDL_events.h
│       ├── SDL_filesystem.h
│       ├── SDL_gamecontroller.h
│       ├── SDL_gesture.h
│       ├── SDL_guid.h
│       ├── SDL.h
│       ├── SDL_haptic.h
│       ├── SDL_hidapi.h
│       ├── SDL_hints.h
│       ├── SDL_joystick.h
│       ├── SDL_keyboard.h
│       ├── SDL_keycode.h
│       ├── SDL_loadso.h
│       ├── SDL_locale.h
│       ├── SDL_log.h
│       ├── SDL_main.h
│       ├── SDL_messagebox.h
│       ├── SDL_metal.h
│       ├── SDL_misc.h
│       ├── SDL_mouse.h
│       ├── SDL_mutex.h
│       ├── SDL_name.h
│       ├── SDL_opengles2_gl2ext.h
│       ├── SDL_opengles2_gl2.h
│       ├── SDL_opengles2_gl2platform.h
│       ├── SDL_opengles2.h
│       ├── SDL_opengles2_khrplatform.h
│       ├── SDL_opengles.h
│       ├── SDL_opengl_glext.h
│       ├── SDL_opengl.h
│       ├── SDL_pixels.h
│       ├── SDL_platform.h
│       ├── SDL_power.h
│       ├── SDL_quit.h
│       ├── SDL_rect.h
│       ├── SDL_render.h
│       ├── SDL_revision.h
│       ├── SDL_rwops.h
│       ├── SDL_scancode.h
│       ├── SDL_sensor.h
│       ├── SDL_shape.h
│       ├── SDL_stdinc.h
│       ├── SDL_surface.h
│       ├── SDL_system.h
│       ├── SDL_syswm.h
│       ├── SDL_test_assert.h
│       ├── SDL_test_common.h
│       ├── SDL_test_compare.h
│       ├── SDL_test_crc32.h
│       ├── SDL_test_font.h
│       ├── SDL_test_fuzzer.h
│       ├── SDL_test.h
│       ├── SDL_test_harness.h
│       ├── SDL_test_images.h
│       ├── SDL_test_log.h
│       ├── SDL_test_md5.h
│       ├── SDL_test_memory.h
│       ├── SDL_test_random.h
│       ├── SDL_thread.h
│       ├── SDL_timer.h
│       ├── SDL_touch.h
│       ├── SDL_types.h
│       ├── SDL_version.h
│       ├── SDL_video.h
│       └── SDL_vulkan.h
├── lib64
│   ├── cmake
│   │   └── SDL2
│   │       ├── SDL2Config.cmake
│   │       ├── SDL2ConfigVersion.cmake
│   │       ├── SDL2mainTargets.cmake
│   │       ├── SDL2mainTargets-debug.cmake
│   │       ├── SDL2-staticTargets.cmake
│   │       ├── SDL2-staticTargets-debug.cmake
│   │       ├── SDL2Targets.cmake
│   │       ├── SDL2Targets-debug.cmake
│   │       ├── SDL2testTargets.cmake
│   │       └── SDL2testTargets-debug.cmake
│   ├── libSDL2-2.0.so -> libSDL2-2.0.so.0
│   ├── libSDL2-2.0.so.0 -> libSDL2-2.0.so.2.90.0
│   ├── libSDL2-2.0.so.2.90.0
│   ├── libSDL2.a
│   ├── libSDL2main.a
│   ├── libSDL2.so -> libSDL2-2.0.so
│   ├── libSDL2test.a
│   └── pkgconfig
│       └── sdl2_compat.pc
└── share
    └── aclocal
        └── sdl2.m4
```

### CODE REVIEW

**Comment by sezero** [comment]
`cmake -DSDL3_INCLUDE_DIRS=xxx` doesn't work. Setting `SDL3_DIR` env var works, though.
```
$ cmake -DCMAKE_BUILD_TYPE=Release -DSDL3_INCLUDE_DIR=/home/ozzie/x3/include/SDL3
[.....]
CMake Warning at CMakeLists.txt:29 (find_package):
  By not providing "FindSDL3.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "SDL3", but
  CMake did not find one.

  Could not find a package configuration file provided by "SDL3" with any of
  the following names:

    SDL3Config.cmake
    sdl3-config.cmake

  Add the installation prefix of "SDL3" to CMAKE_PREFIX_PATH or set
  "SDL3_DIR" to a directory containing one of the above files.  If "SDL3"
  provides a separate development package or SDK, be sure it has been
  installed.
  [.....]
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
/home/ozzie/6/sdl2-compat/SDL3_INCLUDE_DIRS
   used as include directory in directory /home/ozzie/6/sdl2-compat

-- Configuring incomplete, errors occurred!
See also "/home/ozzie/6/sdl2-compat/b/CMakeFiles/CMakeOutput.log".
```

**Comment by sezero** [comment]
I guess `X11_INCLUDE_DIR` addition should be conditional:
```
CMake Error at CMakeLists.txt:74 (include_directories):
  include_directories given empty-string as include directory.
```

**Comment by sezero** [comment]
Mac linkage fails because of that FIXME around line 33:
```
ld: malformed 32-bit x.y.z version number: -current_version
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [libSDL2-2.0.0.dylib] Error 1
make[1]: *** [CMakeFiles/SDL2.dir/all] Error 2
make: *** [all] Error 2
```
We _should_ set dylib versions. SDL2-2.24.2 sets current version to 2401.2
2.26.0 sets its to 2601.0. For simplicity, we can start setting current version
to 2900.0 I guess. As for compatibility version: assuming that newer releases
from SDL2 series won't add new apis, we can set it to 2601.0 yes? (@icculus?)

**Comment by sezero** [comment]
> About versioning of the linux shared library, SDL differentiates between stable and unstable release (even vs uneven number).

Is that supposed to happen???

> How will sdl2-compat handle this?

Sigh.. Badly.. Worst case, we can do what we do with mac version, i.e. define multiple SDL_LIBNAME macros and handle them.

**Comment by slouken** [comment]
I'm not sure what you mean? The library we should load is always libSDL2-2.0.so.0. It will always be a symlink to the current version of SDL 2.x.

**Comment by slouken** [comment]
If you're talking about the one we load, it will always be libSDL3.so.0 on UNIX. The even/odd numbering is in the minor version, which doesn't affect the SONAME.

**Comment by sezero** [comment]
> If you're talking about the one we load, it will always be libSDL3.so.0 on UNIX. The even/odd numbering is in the minor version, which doesn't affect the SONAME.

Well, that's a relief.

**Comment by madebr** [comment]
Yes, we will always load `libSDL3.so.0`, but I was asking about how to version the newly built shared SDL2 library here.

`libSDL2.so.0` will always be created, but how about the full versioned SDL2 libraries?
Will `sdl2-compat` also create `libSDL2-2.0.so.0.9000.X` for release versions and `libsSDL2-2.0.so.0.900X.0` for development versions?

**Comment by sezero** [comment]
AFAICS, with current version, `libSDL2-2.0.so.2.90.0` and other `libSDL2-2.0.so`, `libSDL2-2.0.so.0` and `libSDL2.so` symlinked to it, and `90` incremented as version numbers increase. The problem here is `90` looks like a bad version number considering incremented versions will quickly go past 99.

**Comment by sezero** [comment]
`-DSDL3_INCLUDE_DIRS=xxx` now works.

**Comment by sezero** [comment]
Remaining cmake issues:
- For macOS, we should set `DYLIB_COMPAT_VERSION` and `DYLIB_CURRENT_VERSION` as mentioned above
- `X11_INCLUDE_DIR` addition should be conditional as mentioned above
- `testgles2` and `testgles2_sdf` builds fail miserably when GLES2 isn't installed, such as the case on my system: We should add checks for `HAVE_OPENGLES` and `HAVE_OPENGLES2`, possibly copying the checks from SDL tree, instead of hard-coding them.

**Review by sezero — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,31 +1,68 @@
-# NOTE:  If you installed SDL2 at an unusual place and FindSDL2 fails,
-#        you can define SDL2_INCLUDE_DIR on the cmdline.  For example:
-#        cmake -DSDL2_INCLUDE_DIR=/opt/SDL2/include/SDL2 [other stuff]
+# NOTE:  If you installed SDL3 at an unusual place and SDL3Config fails,
+#        you can define SDL3_INCLUDE_DIRS on the cmdline.  For example:
+#        cmake -DSDL3_INCLUDE_DIRS="/opt/SDL3/include;/opt/SDL3/include/SDL3"
```
Comment (sezero) [inline CMakeLists.txt:3]: Minor nit: The `/SDL3` subdir isn't needed anymore, so the line can simply read:
`#        cmake -DSDL3_INCLUDE_DIRS=/opt/SDL3/include`

**Comment by sezero** [comment]
> AFAICS, with current version, `libSDL2-2.0.so.2.90.0` and 

I was wrong about this: I though the cmake'ry in here did the right thing.
However, SDL's SDL2 branch creates libSDL2-2.0.so.0.2700.0, so, as
@madebr said, we should create libSDL2-2.0.so.0.2900.0

**Pull request merged**